pkg/uefi/nvram.go

```diff
@@ -0,0 +1,458 @@
+	return GUID
+}
+
+func (v *NVar) parseHeader(buf []byte) error {
+	// Read in standard header.
+	r := bytes.NewReader(buf)
+	if err := binary.Read(r, binary.LittleEndian, &v.Header); err != nil {
+		return err
+	}
+	if v.Header.Signature != NVarEntrySignature {
+		return fmt.Errorf("NVAR Signature not found")
+	}
+	if len(buf) < int(v.Header.Size) {
+		return fmt.Errorf("NVAR Size bigger than remaining size")
+	}
+	v.DataOffset = int64(binary.Size(v.Header))
+	return nil
+}
+
+// IsValid tells whether an entry is valid
+func (v *NVar) IsValid() bool {
+	switch v.Type {
+	case LinkNVarEntry, DataNVarEntry, FullNVarEntry:
+		return true
+	default:
+		return false
+	}
+}
+
+func (v *NVar) parseNext() error {
+	var lastVariableFlag uint64
+	if Attributes.ErasePolarity == 0xFF {
+		lastVariableFlag = 0xFFFFFF
+	} else if Attributes.ErasePolarity == 0 {
+		lastVariableFlag = 0
+	} else {
+		return fmt.Errorf("erase polarity not 0x00 or 0xFF, got %#x", Attributes.ErasePolarity)
+	}
+
+	// Add next node information
+	next := Read3Size(v.Header.Next)
+	//fmt.Printf("JVDG: %02x %02x %02x  %06x\n", v.Header.Next[0], v.Header.Next[1], v.Header.Next[2], next)
+	if next != lastVariableFlag {
+		v.Type = LinkNVarEntry
+		v.NextOffset = v.Offset + next
+	}
+	return nil
+}
+
+func (v *NVar) parseExtendedHeader() error {
+	var knownExtDataFormat bool
+	if v.Header.Attributes&NVarEntryExtHeader == 0 {
+		return nil
+	}
+	var extendedHeaderSize uint16
+	r := bytes.NewReader(v.buf)
+	if _, err := r.Seek(-int64(binary.Size(extendedHeaderSize)), io.SeekEnd); err != nil {
+		return err
+	}
+	if err := binary.Read(r, binary.LittleEndian, &extendedHeaderSize); err != nil {
+		return err
+	}
+	// Sanity check extendedHeaderSize < body size
+	bodySize := int64(v.Header.Size) - v.DataOffset
+	if int64(extendedHeaderSize) > bodySize {
+		return fmt.Errorf("extended header size (0x%X) is greater than body size (0x%X)", extendedHeaderSize, bodySize)
+	}
+	v.ExtOffset = int64(v.Header.Size) - int64(extendedHeaderSize)
+	if _, err := r.Seek(v.ExtOffset, io.SeekStart); err != nil {
+		return err
+	}
+	var extAttributes NVarExtAttribute
+	if err := binary.Read(r, binary.LittleEndian, &extAttributes); err != nil {
+		return err
+	}
+	v.ExtAttributes = &extAttributes
+	// Variable with checksum
+	if extAttributes&NVarEntryExtChecksum != 0 {
+		// Get stored checksum
+		var storedChecksum uint8
+		storedChecksum = v.buf[int64(v.Header.Size)-int64(binary.Size(extendedHeaderSize)+binary.Size(storedChecksum))]
+		v.Checksum = &storedChecksum
+		// Recalculate checksum for the variable
+		calculatedChecksum := uint8(0)
+		// [4-5] Include entry size and flags
+		// [ 9 ] Include entry attributes
+		// [10-] Include entry data
+		for i := int64(4); i < int64(v.Header.Size); i++ {
+			calculatedChecksum += v.buf[i]
+			if i == 5 {
+				i += 3 // Skip Next
+			}
+		}
+		if calculatedChecksum != 0 {
+			calculatedChecksum = -calculatedChecksum
+			v.ExpectedChecksum = &calculatedChecksum
```

Comment:
Not really, the value will be present in JSON only if it's not valid... more an information.

Reply:
I'm ok with this for now.

If you look in the rest of the package, the checksum validation for the other fields are done in the validate visitor, and not done while parsing. This was done a long time ago because we had some rom which failed checksums while parsing. I'm now no longer sure that's the right decision, and maybe we should do aggressive checksum checking during parsing and emit error messages or warnings if we see them, because if the file header is corrupt, who's to know if the size we read off the header is actually correct?

Long term I would rather remove the checksum calculations from validate.go, and instead emit warnings or errors. Perhaps it should default to an error, and take a flag to turn those to warnings.

Reply:
The validate visitor still currently fail on my winterfell images...
So while I agree with your long term goal, some more work will probably be needed before that... I guess some stuff are parsed while some other flags would have told upfront that the portion is invalid and should not be parsed ... (really just a guess)

Reply:
I put that on my TODO list